File: app/src/main/java/ir/ghararemaghzha/game/fragments/BuyHistoryFragment.java
package ir.ghararemaghzha.game.fragments;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.textview.MaterialTextView;

import ir.ghararemaghzha.game.R;
import ir.ghararemaghzha.game.adapters.BuyHistoryAdapter;
import ir.ghararemaghzha.game.classes.MySharedPreference;
import ir.ghararemaghzha.game.classes.Utils;
import ir.ghararemaghzha.game.data.RetrofitClient;
import ir.ghararemaghzha.game.models.BuyHistoryResponse;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class BuyHistoryFragment extends Fragment {
    private Context context;
    private FragmentActivity activity;
    private RecyclerView recyclerView;
    private BuyHistoryAdapter adapter;
    private ConstraintLayout loading;
    private ConstraintLayout empty;

    public BuyHistoryFragment() {
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_buy_history, container, false);
        context = getContext();
        activity = getActivity();
        init(v);
        return v;
    }


    private void init(View v) {
        ((MaterialTextView) activity.findViewById(R.id.toolbar_title)).setText(R.string.buyhistory_title);

        loading = v.findViewById(R.id.buyhistory_loading);
        empty = v.findViewById(R.id.buyhistory_empty);
        recyclerView = v.findViewById(R.id.buyhistory_recycler);
        recyclerView.setLayoutManager(new LinearLayoutManager(context));
        getData();

    }

    private void getData() {
        String number = MySharedPreference.Companion.getInstance(context).getNumber();
        String token = MySharedPreference.Companion.getInstance(context).getAccessToken();
        if (number.isEmpty() || token.isEmpty()) {
            Utils.logout(activity, true);
            return;
        }
        loading.setVisibility(View.VISIBLE);
        RetrofitClient.Companion.getInstance().getApi()
                .getBuyHistory("Bearer " + token, number)
                .enqueue(new Callback<BuyHistoryResponse>() {
                    @Override
                    public void onResponse(@NonNull Call<BuyHistoryResponse> call, @NonNull Response<BuyHistoryResponse> response) {
                        loading.setVisibility(View.GONE);
                        if (response.isSuccessful() && response.body() != null && response.body().getResult().equals("success")) {
                            if (response.body().getMessage().equals("all history")) {
                                adapter = new BuyHistoryAdapter(context, response.body().getData());
                                recyclerView.setAdapter(adapter);
                            } else {
                                empty.setVisibility(View.VISIBLE);
                            }
                        } else if (response.code() == 401) {
                            Utils.logout(activity, true);
                        } else {
                            Toast.makeText(context, getString(R.string.general_error), Toast.LENGTH_SHORT).show();
                            empty.setVisibility(View.VISIBLE);
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<BuyHistoryResponse> call, @NonNull Throwable t) {
                        loading.setVisibility(View.GONE);
                        Toast.makeText(context, getString(R.string.general_error), Toast.LENGTH_SHORT).show();
                        empty.setVisibility(View.VISIBLE);
                    }
                });
    }
}

File: app/src/main/java/ir/ghararemaghzha/game/fragments/BuyHistoryFragment.kt
package ir.ghararemaghzha.game.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.textview.MaterialTextView
import ir.ghararemaghzha.game.R
import ir.ghararemaghzha.game.adapters.BuyHistoryAdapter
import ir.ghararemaghzha.game.classes.MySharedPreference
import ir.ghararemaghzha.game.classes.Utils
import ir.ghararemaghzha.game.classes.Utils.Companion.showInternetError
import ir.ghararemaghzha.game.data.Resource
import ir.ghararemaghzha.game.databinding.FragmentBuyHistoryBinding
import ir.ghararemaghzha.game.viewmodels.BuyHistoryViewModel

class BuyHistoryFragment : BaseFragment<BuyHistoryViewModel, FragmentBuyHistoryBinding>() {

    private lateinit var adapter: BuyHistoryAdapter
    private lateinit var number: String
    private lateinit var token: String


    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        init()

        viewModel.buyHistoryResponse.observe(viewLifecycleOwner, { res ->
            when (res) {
                is Resource.Success -> {

                    if (res.value.result == "success" && res.value.message == "all history") {
                        val data = res.value.data
                        b.buyhistoryLoading.visibility = View.GONE
                        adapter = BuyHistoryAdapter(requireContext(), data)
                        b.buyhistoryRecycler.adapter = adapter

                    } else {
                        b.buyhistoryLoading.visibility = View.GONE
                        b.buyhistoryEmpty.visibility = View.VISIBLE
                    }

                }
                is Resource.Failure -> {

                    if (res.isNetworkError) {
                        showInternetError(requireContext()) { viewModel.getBuyHistory("Bearer $token", number) }
                        Toast.makeText(context, getString(R.string.general_error), Toast.LENGTH_SHORT).show()

                    } else if (res.errorCode == 401) {
                        b.buyhistoryLoading.visibility = View.GONE
                        Utils.logout(requireActivity(), true)
                    }
                }
                is Resource.Loading -> {
                }
            }
        })
    }


    private fun init() {
        requireActivity().findViewById<MaterialTextView>(R.id.toolbar_title).setText(R.string.buyhistory_title)

        b.buyhistoryRecycler.layoutManager = LinearLayoutManager(context)
        b.buyhistoryLoading.visibility = View.VISIBLE

        number = MySharedPreference.getInstance(requireContext()).getNumber()
        token = MySharedPreference.getInstance(requireContext()).getAccessToken()
        if (number.isEmpty() || token.isEmpty()) {
            Utils.logout(requireActivity(), true)
            return
        }

        viewModel.getBuyHistory("Bearer $token", number)

    }

    override fun getViewModel() = BuyHistoryViewModel::class.java

    override fun getFragmentBinding(inflater: LayoutInflater, container: ViewGroup?) = FragmentBuyHistoryBinding.inflate(inflater, container, false)
}